Utils/XDrive.hpp

```diff
@@ -57,8 +57,130 @@ class XDriveRobot {
+    double brxdriveVelocity (double speed, double tgtHeading, double wheelAngle = 45) {
+        return speed * sin(degree2arc(90-wheelAngle+tgtHeading));
+    }
+
+    double blxdriveVelocity (double speed, double tgtHeading, double wheelAngle = 45) {
+        return speed * sin(degree2arc(90-wheelAngle-tgtHeading));
+    }
+
+    Coord goStraight(double dist, vex::directionType dt, double tgtHeading, double originalSpeed, double kp = 0.02, vex::brakeType bt = brake, Coord srcLoc = Coord(0.0, 0.0));
+
 };  // class XDriveRobot
 
+Coord XDriveRobot::goStraight(double dist, vex::directionType dt, double tgtHeading, double originalSpeed, double kp, vex::brakeType bt, Coord srcLoc) {
+  Coord currLoc = srcLoc;
+
+  resetRot();
+
+  // orientation of the wheels, default = 45
+  double wheelAngle = 45;
+  double wheelArc = degree2arc(45);
+
+  // target orientation that the robot should cont. to travel at
+  // tgtHeading is a parameter
+  double targetArc = degree2arc(tgtHeading);
+  
+  // tracking distance
+  double distToGo = dist; // distance more to travel
+  double distTravelled = dist - distToGo; // distance already traveled
+  double prevRot = backleftdrive.rotation(vex::rotationUnits::deg); // amount of rotation at the last run through
+  double changedRot = backleftdrive.rotation(vex::rotationUnits::deg) - prevRot; // rotations passed since last loop
+
+  // speed
+  double finalSpeed = 10; // capping speed
+  double speed = originalSpeed; // max speed
+  double const adaptiveInterval = 10; // slow down/speed up interval
+
+  
+  // location tracking on the coordinate plane
+  double dx = 0; // change in x coordinate since last loop
+  double dy = 0; // change in y coordinate since last loop
+
+  double xSpeed = originalSpeed * cos(targetArc);
+  double ySpeed = originalSpeed * sin(targetArc);
+
+  double blspeed = xSpeed - ySpeed;
+  double brspeed = -xSpeed - ySpeed;
+  double flspeed = -xSpeed + ySpeed;
+  double frspeed = xSpeed + ySpeed;
+
+  while (distToGo > 0) {
+
+    double frsign = 1;
+    double flsign = -1;
+    double blsign = -1;
+    double brsign = 1;
+
+    // update other variables
+
+    double headingError = inertialSensor.heading() - tgtHeading; // updates the error in heading
+
+    // adjusts the heading error so that it is between -180 and 180
+    if (headingError < -180) headingError = headingError + 360;
+    if (headingError > 180) headingError = headingError - 360;
+
+    // caps headingError so that adjustments are not too dramatic
+    if (headingError < -15) headingError = -15;
+    if (headingError > 15) headingError = 15;
+
+    // by default, speed with no adjustments is as such
+    speed = originalSpeed;
+
+    if (distTravelled < adaptiveInterval) speed = originalSpeed * distTravelled / adaptiveInterval;
+    if (distToGo < adaptiveInterval) speed = originalSpeed * (1 - (adaptiveInterval - distToGo) / adaptiveInterval);
+
+    if (speed < finalSpeed) speed = finalSpeed;
+
+    if (headingError > 0) {
```

Comment:
use sign() function you implemented to consolidate if..else cases into one case. To make the line shorter (and easier to read), you can define kpHeadingErr = kp * headingError then use it.